File: sdk/go-holons/pkg/observability/observability_test.go
package observability

import (
	"context"
	"os"
	"sync"
	"testing"
	"time"
)

func TestParseOPOBS_Basic(t *testing.T) {
	tests := []struct {
		in   string
		want map[Family]bool
	}{
		{"", map[Family]bool{}},
		{"logs", map[Family]bool{FamilyLogs: true}},
		{"logs,metrics", map[Family]bool{FamilyLogs: true, FamilyMetrics: true}},
		{"all", map[Family]bool{FamilyLogs: true, FamilyMetrics: true, FamilyEvents: true, FamilyProm: true}},
		{"all,otel", map[Family]bool{FamilyLogs: true, FamilyMetrics: true, FamilyEvents: true, FamilyProm: true}}, // otel silently dropped in v1
		{"unknown", map[Family]bool{}},
	}
	for _, tc := range tests {
		got := parseOPOBS(tc.in)
		if len(got) != len(tc.want) {
			t.Errorf("parseOPOBS(%q) len=%d, want %d; got=%v", tc.in, len(got), len(tc.want), got)
			continue
		}
		for k, v := range tc.want {
			if got[k] != v {
				t.Errorf("parseOPOBS(%q)[%v]=%v, want %v", tc.in, k, got[k], v)
			}
		}
	}
}

func TestCheckEnv_OtelRejected(t *testing.T) {
	t.Setenv("OP_OBS", "logs,otel")
	err := CheckEnv()
	if err == nil {
		t.Fatal("expected error for otel in v1")
	}
	if ite, ok := err.(*InvalidTokenError); !ok || ite.Token != "otel" {
		t.Fatalf("expected InvalidTokenError{Token:otel}, got %v", err)
	}
}

func TestCheckEnv_UnknownRejected(t *testing.T) {
	t.Setenv("OP_OBS", "bogus")
	err := CheckEnv()
	if err == nil {
		t.Fatal("expected error for unknown token")
	}
}

func TestCheckEnv_Valid(t *testing.T) {
	t.Setenv("OP_OBS", "logs,metrics,events,prom,all")
	if err := CheckEnv(); err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestConfigure_Disabled_ZeroCost(t *testing.T) {
	Reset()
	os.Unsetenv("OP_OBS")
	obs := Configure(Config{Slug: "test"})
	defer obs.Close()

	if obs.Enabled(FamilyLogs) || obs.Enabled(FamilyMetrics) || obs.Enabled(FamilyEvents) {
		t.Error("no family should be enabled with empty OP_OBS")
	}
	l := obs.Logger("test")
	l.Info("noop", "k", "v") // should be a no-op

	c := obs.Counter("test_total", "", nil)
	if c != nil {
		t.Error("counter should be nil when metrics disabled")
	}
	c.Add(1) // nil-safe
	c.Inc()
	if c.Value() != 0 {
		t.Error("nil counter must read 0")
	}
}

func TestConfigure_LogsFamily(t *testing.T) {
	Reset()
	t.Setenv("OP_OBS", "logs")
	obs := Configure(Config{Slug: "g", InstanceUID: "uid"})
	defer obs.Close()

	if !obs.Enabled(FamilyLogs) {
		t.Fatal("logs family not enabled")
	}
	l := obs.Logger("recipe.runner")
	l.Info("recipe started", "name", "greeter")
	l.Warn("retry budget low", "remaining", 3)

	entries := obs.LogRing().Drain()
	if len(entries) != 2 {
		t.Fatalf("expected 2 entries, got %d", len(entries))
	}
	if entries[0].Message != "recipe started" || entries[1].Message != "retry budget low" {
		t.Errorf("unexpected messages: %+v", entries)
	}
	if entries[0].Slug != "g" || entries[0].InstanceUID != "uid" {
		t.Errorf("well-known fields not injected: %+v", entries[0])
	}
	if entries[0].Fields["name"] != "greeter" {
		t.Errorf("user field missing: %+v", entries[0].Fields)
	}
}

func TestLogger_LevelFiltering(t *testing.T) {
	Reset()
	t.Setenv("OP_OBS", "logs")
	obs := Configure(Config{Slug: "g", DefaultLogLevel: LevelWarn})
	defer obs.Close()

	l := obs.Logger("filter")
	l.Info("should be dropped")
	l.Warn("should pass")
	l.Error("should pass")

	if n := obs.LogRing().Len(); n != 2 {
		t.Fatalf("expected 2 entries at WARN+, got %d", n)
	}
}

func TestLogger_RedactFields(t *testing.T) {
	Reset()
	t.Setenv("OP_OBS", "logs")
	obs := Configure(Config{Slug: "g", RedactedFields: []string{"password", "api_key"}})
	defer obs.Close()

	obs.Logger("login").Info("authenticated", "user", "bob", "password", "secret", "api_key", "abc123")

	e := obs.LogRing().Drain()[0]
	if e.Fields["user"] != "bob" {
		t.Errorf("user field unexpectedly altered: %v", e.Fields["user"])
	}
	if e.Fields["password"] != "<redacted>" || e.Fields["api_key"] != "<redacted>" {
		t.Errorf("redaction failed: %v", e.Fields)
	}
}

func TestRing_PushEvictsOldest(t *testing.T) {
	r := NewLogRing(3)
	for i := 0; i < 5; i++ {
		r.Push(LogEntry{Message: string(rune('a' + i))})
	}
	entries := r.Drain()
	if len(entries) != 3 {
		t.Fatalf("expected 3, got %d", len(entries))
	}
	if entries[0].Message != "c" || entries[1].Message != "d" || entries[2].Message != "e" {
		t.Errorf("wrong order: %+v", entries)
	}
}

func TestRing_Watch(t *testing.T) {
	r := NewLogRing(10)
	ch, stop := r.Watch(4)
	defer stop()

	var wg sync.WaitGroup
	wg.Add(1)
	received := make([]LogEntry, 0, 3)
	go func() {
		defer wg.Done()
		for i := 0; i < 3; i++ {
			e := <-ch
			received = append(received, e)
		}
	}()

	r.Push(LogEntry{Message: "1"})
	r.Push(LogEntry{Message: "2"})
	r.Push(LogEntry{Message: "3"})

	wg.Wait()
	if len(received) != 3 {
		t.Fatalf("expected 3 live entries, got %d", len(received))
	}
}

func TestCounter_Atomic(t *testing.T) {
	Reset()
	t.Setenv("OP_OBS", "metrics")
	obs := Configure(Config{Slug: "g"})
	defer obs.Close()

	c := obs.Counter("test_total", "", nil)
	var wg sync.WaitGroup
	for i := 0; i < 100; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for j := 0; j < 100; j++ {
				c.Inc()
			}
		}()
	}
	wg.Wait()
	if c.Value() != 10000 {
		t.Errorf("expected 10000, got %d", c.Value())
	}
}

func TestHistogram_Percentile(t *testing.T) {
	Reset()
	t.Setenv("OP_OBS", "metrics")
	obs := Configure(Config{Slug: "g"})
	defer obs.Close()

	h := obs.Histogram("latency_seconds", "", nil, []float64{1e-3, 1e-2, 1e-1, 1.0})
	for i := 0; i < 900; i++ {
		h.Observe(0.5e-3) // below smallest bucket (0.001); counts in bucket[0]
	}
	for i := 0; i < 100; i++ {
		h.Observe(0.5) // counts in bucket[3] (1.0)
	}
	snap := h.Snapshot()
	p50 := snap.Quantile(0.50)
	p99 := snap.Quantile(0.99)
	if p50 != 1e-3 {
		t.Errorf("expected p50=1e-3 (first bucket), got %v", p50)
	}
	if p99 != 1.0 {
		t.Errorf("expected p99=1.0, got %v", p99)
	}
}

func TestEventBus_FanOut(t *testing.T) {
	Reset()
	t.Setenv("OP_OBS", "events")
	obs := Configure(Config{Slug: "g", InstanceUID: "uid"})
	defer obs.Close()

	bus := obs.EventBus()
	ch1, stop1 := bus.Watch(8)
	defer stop1()
	ch2, stop2 := bus.Watch(8)
	defer stop2()

	obs.Emit(context.Background(), EventInstanceReady, map[string]string{"listener": "stdio://"})
	time.Sleep(5 * time.Millisecond) // let subscribers receive

	select {
	case e := <-ch1:
		if e.Type != EventInstanceReady {
			t.Errorf("ch1: unexpected event: %+v", e)
		}
	default:
		t.Error("ch1 did not receive event")
	}
	select {
	case e := <-ch2:
		if e.Type != EventInstanceReady {
			t.Errorf("ch2: unexpected event: %+v", e)
		}
	default:
		t.Error("ch2 did not receive event")
	}

	// Drain returns the full buffer history.
	ev := bus.Drain()
	if len(ev) != 1 {
		t.Errorf("expected 1 event in drain, got %d", len(ev))
	}
}

func TestChain_AppendAndEnrich(t *testing.T) {
	// Wire chain: start with [], direct-child relay appends.
	c1 := AppendDirectChild(nil, "gabriel-greeting-rust", "1c2d")
	if len(c1) != 1 || c1[0].Slug != "gabriel-greeting-rust" {
		t.Fatalf("append direct child failed: %+v", c1)
	}

	// Multilog enrichment appends stream source.
	c2 := EnrichForMultilog(c1, "gabriel-greeting-go", "ea34")
	if len(c2) != 2 || c2[0].Slug != "gabriel-greeting-rust" || c2[1].Slug != "gabriel-greeting-go" {
		t.Fatalf("multilog enrichment failed: %+v", c2)
	}
	// Original wire chain must be unchanged.
	if len(c1) != 1 {
		t.Fatalf("wire chain mutated: %+v", c1)
	}
}

func TestIsOrganismRoot(t *testing.T) {
	Reset()
	os.Unsetenv("OP_OBS")
	obs := Configure(Config{Slug: "g", InstanceUID: "x", OrganismUID: "x"})
	defer obs.Close()
	if !obs.IsOrganismRoot() {
		t.Error("expected root when OP_ORGANISM_UID == OP_INSTANCE_UID")
	}
	Reset()
	obs2 := Configure(Config{Slug: "g", InstanceUID: "x", OrganismUID: "y"})
	defer obs2.Close()
	if obs2.IsOrganismRoot() {
		t.Error("expected not-root when uids differ")
	}
}

func TestCurrent_SafeWhenUnset(t *testing.T) {
	Reset()
	c := Current()
	if c == nil {
		t.Fatal("Current should never return nil; disabled stub expected")
	}
	c.Logger("anything").Info("no-op call path")
	c.Emit(nil, EventInstanceReady, nil)
}

var _ = time.Second

File: sdk/node-holons/tests/observability.test.js
'use strict';

const test = require('node:test');
const assert = require('node:assert/strict');
const EventEmitter = require('node:events');
const fs = require('node:fs');
const net = require('node:net');
const os = require('node:os');
const path = require('node:path');
const grpc = require('@grpc/grpc-js');

const obs = require('../src/observability');
const describe = require('../src/describe');
const serve = require('../src/serve');
const transport = require('../src/transport');
const observabilityWire = require('../src/gen/holons/v1/observability');
const { sampleStaticDescribeResponse } = require('./helpers/static_describe');

test.beforeEach(() => {
  obs.reset();
  delete process.env.OP_OBS;
  delete process.env.OP_RUN_DIR;
  delete process.env.OP_INSTANCE_UID;
  describe.useStaticResponse(null);
});

test('parseOpObs basic', () => {
  const cases = [
    ['', new Set()],
    ['logs', new Set([obs.Family.LOGS])],
    ['logs,metrics', new Set([obs.Family.LOGS, obs.Family.METRICS])],
    ['all', new Set([obs.Family.LOGS, obs.Family.METRICS, obs.Family.EVENTS, obs.Family.PROM])],
  ];
  for (const [input, want] of cases) {
    const got = obs.parseOpObs(input);
    assert.equal(got.size, want.size, input);
    for (const f of want) assert.ok(got.has(f), `${input}: missing ${f}`);
  }
  for (const input of ['all,otel', 'all,sessions', 'unknown']) {
    assert.throws(() => obs.parseOpObs(input), obs.InvalidTokenError);
  }
});

test('checkEnv rejects unknown OP_OBS tokens', () => {
  process.env.OP_OBS = 'logs,otel';
  assert.throws(() => obs.checkEnv(), obs.InvalidTokenError);

  process.env.OP_OBS = 'bogus';
  assert.throws(() => obs.checkEnv(), obs.InvalidTokenError);

  process.env.OP_OBS = 'logs,sessions';
  assert.throws(() => obs.checkEnv(), obs.InvalidTokenError);

  process.env.OP_OBS = 'logs,metrics,events,prom,all';
  obs.checkEnv();
});

test('disabled is no-op', () => {
  const o = obs.configure({ slug: 't' });
  assert.equal(o.enabled(obs.Family.LOGS), false);
  o.logger('x').info('drop', { k: 'v' });
  assert.equal(o.counter('t_total'), null);
});

test('logs ring emits LogRecord with typed AnyValue attributes', () => {
  process.env.OP_OBS = 'logs';
  const o = obs.configure({ slug: 'g', instanceUid: 'uid', redactedFields: ['password'] });
  const l = o.logger('r');
  l.info('hello', {
    who: 'bob',
    ok: true,
    count: 3,
    ratio: 1.5,
    object: { value: 1 },
    password: 'secret',
  });
  const entries = o.logRing.drain();
  assert.equal(entries.length, 1);
  const record = entries[0];
  assert.equal(obs.bodyString(record), 'hello');
  assert.equal(record.severity_number, obs.Level.INFO);
  assert.equal(record.severity_text, 'INFO');
  assert.equal(obs.stringAttribute(record, obs.Attr.HOLONS_SLUG), 'g');
  assert.equal(obs.stringAttribute(record, obs.Attr.SERVICE_NAME), 'g');
  assert.equal(obs.stringAttribute(record, obs.Attr.HOLONS_INSTANCE_UID), 'uid');
  assert.equal(obs.stringAttribute(record, obs.Attr.SERVICE_INSTANCE_ID), 'uid');
  assert.equal(obs.stringAttribute(record, 'who'), 'bob');
  assert.deepEqual(attr(record, 'ok').value, { bool_value: true });
  assert.deepEqual(attr(record, 'count').value, { int_value: 3 });
  assert.deepEqual(attr(record, 'ratio').value, { double_value: 1.5 });
  assert.deepEqual(attr(record, 'object').value, { string_value: '[object Object]' });
  assert.deepEqual(attr(record, 'password').value, { string_value: '<redacted>' });
});

test('logs include current session context attributes when available', () => {
  process.env.OP_OBS = 'logs';
  const o = obs.configure({ slug: 'g', instanceUid: 'uid' });
  obs.withSessionContext({ sessionId: 'session-1', rpcMethod: '/greeting.v1.GreetingService/SayHello' }, () => {
    o.logger('r').info('inside session');
  });
  const record = o.logRing.drain()[0];
  assert.equal(obs.stringAttribute(record, obs.Attr.HOLONS_SESSION_ID), 'session-1');
  assert.equal(obs.stringAttribute(record, obs.Attr.RPC_METHOD), '/greeting.v1.GreetingService/SayHello');
});

test('counter increments and histogram percentile', () => {
  process.env.OP_OBS = 'metrics';
  const o = obs.configure({ slug: 'g' });
  const c = o.counter('t_total');
  for (let i = 0; i < 1000; i++) c.inc();
  assert.equal(c.value(), 1000);

  const h = o.histogram('lat_s', '', {}, [1e-3, 1e-2, 1e-1, 1.0]);
  for (let i = 0; i < 900; i++) h.observe(0.5e-3);
  for (let i = 0; i < 100; i++) h.observe(0.5);
  const snap = h.snapshot();
  assert.equal(obs.Histogram.quantile(snap, 0.5), 1e-3);
  assert.equal(obs.Histogram.quantile(snap, 0.99), 1.0);
});

test('event bus fan-out', () => {
  process.env.OP_OBS = 'events';
  const o = obs.configure({ slug: 'g', instanceUid: 'uid' });
  const received = [];
  o.eventBus.subscribe((e) => received.push(e));
  o.emit(obs.EventName.INSTANCE_READY, { listener: 'stdio://' });
  assert.equal(received.length, 1);
  assert.equal(received[0].event_name, obs.EventName.INSTANCE_READY);
  assert.equal(obs.stringAttribute(received[0], 'listener'), 'stdio://');
});

test('chain append and enrichment', () => {
  const c1 = obs.appendDirectChild([], 'gabriel-greeting-rust', '1c2d');
  assert.equal(c1.length, 1);
  assert.equal(c1[0], 'gabriel-greeting-rust');
  const c2 = obs.enrichForMultilog(c1, 'gabriel-greeting-go', 'ea34');
  assert.equal(c2.length, 2);
  assert.equal(c2[1], 'gabriel-greeting-go');
  assert.equal(c1.length, 1); // original unchanged
});

test('isOrganismRoot', () => {
  const o1 = obs.configure({ slug: 'g', instanceUid: 'x', organismUid: 'x' });
  assert.equal(o1.isOrganismRoot(), true);
  obs.reset();
  const o2 = obs.configure({ slug: 'g', instanceUid: 'x', organismUid: 'y' });
  assert.equal(o2.isOrganismRoot(), false);
});

test('current never returns null', () => {
  const c = obs.current();
  assert.ok(c);
  c.logger('x').info('safe'); // must not throw
});

test('run dir derives from registry root', () => {
  process.env.OP_OBS = 'logs';
  const root = fs.mkdtempSync(path.join(os.tmpdir(), 'js-obs-root-'));
  const o = obs.configure({ slug: 'gabriel', instanceUid: 'uid-1', runDir: root });
  assert.equal(o.cfg.runDir, path.join(root, 'gabriel', 'uid-1'));
});

test('disk writers and meta.json use instance run dir', () => {
  process.env.OP_OBS = 'logs,events';
  const root = fs.mkdtempSync(path.join(os.tmpdir(), 'js-obs-disk-'));
  const o = obs.configure({ slug: 'gabriel', instanceUid: 'uid-1', runDir: root });
  obs.enableDiskWriters(o.cfg.runDir);
  o.logger('test').info('ready', { port: 123 });
  o.emit(obs.EventName.INSTANCE_READY, { listener: 'tcp://127.0.0.1:123' });
  obs.writeMetaJson(o.cfg.runDir, {
    slug: 'gabriel',
    uid: 'uid-1',
    pid: 42,
    started_at: new Date(1000).toISOString(),
    mode: 'persistent',
    transport: 'tcp',
    address: 'tcp://127.0.0.1:123',
    log_path: path.join(o.cfg.runDir, 'stdout.log'),
  });

  assert.match(fs.readFileSync(path.join(root, 'gabriel', 'uid-1', 'stdout.log'), 'utf8'), /ready/);
  assert.match(fs.readFileSync(path.join(root, 'gabriel', 'uid-1', 'events.jsonl'), 'utf8'), /instance\.ready/);
  const meta = obs.readMetaJson(o.cfg.runDir);
  assert.equal(meta.slug, 'gabriel');
  assert.equal(meta.uid, 'uid-1');
  assert.equal(meta.address, 'tcp://127.0.0.1:123');
});

test('HolonObservability handlers replay logs, metrics, and events', async () => {
  process.env.OP_OBS = 'logs,metrics,events';
  const o = obs.configure({ slug: 'gabriel', instanceUid: 'uid-1' });
  o.logger('test').info('hello');
  o.counter('requests_total', 'requests').inc();
  o.emit(obs.EventName.INSTANCE_READY, { listener: 'stdio://' });

  const handlers = obs.makeHolonObservabilityHandlers(o);
  const logs = new FakeStreamCall({ follow: false });
  handlers.Logs(logs);
  assert.deepEqual(logs.writes.map((entry) => obs.bodyString(entry)), ['hello']);
  assert.equal(logs.ended, true);

  const metrics = new FakeStreamCall({});
  handlers.Metrics(metrics);
  const requests = metrics.writes.find((metric) => metric.name === 'requests_total');
  assert.ok(requests);
  assert.equal(requests.sum.is_monotonic, true);
  assert.equal(requests.sum.aggregation_temporality, 'AGGREGATION_TEMPORALITY_CUMULATIVE');
  assert.equal(requests.sum.data_points[0].as_int, 1);
  assert.equal(obs.stringAttribute(requests.sum.data_points[0].attributes, obs.Attr.HOLONS_SLUG), 'gabriel');
  assert.equal(obs.stringAttribute(requests.sum.data_points[0].attributes, obs.Attr.HOLONS_INSTANCE_UID), 'uid-1');

  const events = new FakeStreamCall({ follow: false });
  handlers.Events(events);
  assert.deepEqual(events.writes.map((event) => event.event_name), [obs.EventName.INSTANCE_READY]);
});

test('metric oneofs encode counters, gauges, and histograms', () => {
  process.env.OP_OBS = 'metrics';
  const o = obs.configure({ slug: 'gabriel', instanceUid: 'uid-1' });
  o.counter('requests_total', 'requests').inc(2);
  o.gauge('temperature_celsius', 'temperature').set(21.5);
  const histogram = o.histogram('latency_seconds', 'latency', {}, [0.1, 1.0]);
  histogram.observe(0.05);
  histogram.observe(2.0);

  const metrics = obs.toProtoMetrics(o.registry.snapshot(), o.cfg, o.startUnixNano);
  const counter = metrics.find((metric) => metric.name === 'requests_total');
  const gauge = metrics.find((metric) => metric.name === 'temperature_celsius');
  const hist = metrics.find((metric) => metric.name === 'latency_seconds');
  assert.equal(counter.sum.data_points[0].as_int, 2);
  assert.equal(counter.sum.is_monotonic, true);
  assert.equal(gauge.gauge.data_points[0].as_double, 21.5);
  assert.deepEqual(hist.histogram.data_points[0].bucket_counts, [1, 0, 1]);
  assert.deepEqual(hist.histogram.data_points[0].explicit_bounds, [0.1, 1.0]);
  assert.equal(hist.histogram.data_points[0].count, 2);
  assert.equal(hist.histogram.data_points[0].sum, 2.05);
  assert.equal(hist.histogram.data_points[0].min, 0.05);
  assert.equal(hist.histogram.data_points[0].max, 2.0);
});

test('Logs(follow=true) replays ring before live entries', () => {
  process.env.OP_OBS = 'logs';
  const o = obs.configure({ slug: 'gabriel', instanceUid: 'uid-1' });
  o.logger('test').info('before');

  const handlers = obs.makeHolonObservabilityHandlers(o);
  const logs = new FakeStreamCall({ follow: true });
  handlers.Logs(logs);
  assert.deepEqual(logs.writes.map((entry) => obs.bodyString(entry)), ['before']);

  o.logger('test').info('after');
  assert.deepEqual(logs.writes.map((entry) => obs.bodyString(entry)), ['before', 'after']);
  logs.emit('close');
});

test('Events(follow=true) replays ring before live entries', () => {
  process.env.OP_OBS = 'events';
  const o = obs.configure({ slug: 'gabriel', instanceUid: 'uid-1' });
  o.emit(obs.EventName.INSTANCE_READY, { listener: 'stdio://' });

  const handlers = obs.makeHolonObservabilityHandlers(o);
  const events = new FakeStreamCall({ follow: true });
  handlers.Events(events);
  assert.deepEqual(events.writes.map((event) => event.event_name), [obs.EventName.INSTANCE_READY]);

  o.emit(obs.EventName.CONFIG_RELOADED, { source: 'test' });
  assert.deepEqual(events.writes.map((event) => event.event_name), [obs.EventName.INSTANCE_READY, obs.EventName.CONFIG_RELOADED]);
  events.emit('close');
});

test('private log and event emissions remain local but not follow-streamed', () => {
  process.env.OP_OBS = 'logs,events';
  const o = obs.configure({ slug: 'gabriel', instanceUid: 'uid-1' });
  o.logger('test').private().info('secret');
  o.logger('test').info('public');
  o.emitPrivate(obs.EventName.CONFIG_RELOADED, { source: 'secret' });
  o.emit(obs.EventName.INSTANCE_READY, { listener: 'stdio://' });

  const handlers = obs.makeHolonObservabilityHandlers(o);
  const logs = new FakeStreamCall({ follow: true });
  handlers.Logs(logs);
  assert.deepEqual(o.logRing.drain().map((entry) => obs.bodyString(entry)), ['secret', 'public']);
  assert.deepEqual(logs.writes.map((entry) => obs.bodyString(entry)), ['public']);
  logs.emit('close');

  const events = new FakeStreamCall({ follow: true });
  handlers.Events(events);
  assert.equal(o.eventBus.drain().length, 2);
  assert.deepEqual(events.writes.map((event) => event.event_name), [obs.EventName.INSTANCE_READY]);
  events.emit('close');
});

test('Prometheus exposition includes injected runtime labels', async () => {
  process.env.OP_OBS = 'all';
  const o = obs.configure({ slug: 'node-prom', instanceUid: 'uid-prom' });
  o.counter('cascade_ticks_total', 'Ticks received by this cascade node.', { responder_uid: 'uid-prom' }).inc();

  const text = obs.toPrometheusText(o);
  assert.match(text, /# HELP cascade_ticks_total Ticks received by this cascade node\./);
  assert.ok(text.includes('cascade_ticks_total{instance_uid="uid-prom",responder_uid="uid-prom",slug="node-prom"} 1'));

  const prom = new obs.PromServer('127.0.0.1:0');
  try {
    const addr = await prom.start();
    const body = await httpGet(addr);
    assert.match(body, /cascade_ticks_total/);
  } finally {
    await prom.close();
  }
});

test('MemberRelay forwards logs and events with child chain hop', () => {
  process.env.OP_OBS = 'logs,events';
  const parent = obs.configure({ slug: 'parent-node', instanceUid: 'parent-uid' });
  const logStream = new EventEmitter();
  const eventStream = new EventEmitter();
  logStream.cancel = () => {};
  eventStream.cancel = () => {};
  const relay = new obs.MemberRelay({
    childSlug: 'child-node',
    childUid: 'child-uid',
    client: {
      Logs() { return logStream; },
      Events() { return eventStream; },
    },
    observability: parent,
    retryDelayMs: 10,
  });
  relay.start();
  logStream.emit('data', {
    time_unix_nano: '1000000000',
    observed_time_unix_nano: '1000000000',
    severity_number: obs.Level.INFO,
    severity_text: 'INFO',
    body: { string_value: 'relay-log' },
    attributes: [
      obs.keyValue(obs.Attr.HOLONS_SLUG, 'child-node'),
      obs.keyValue(obs.Attr.HOLONS_INSTANCE_UID, 'child-uid'),
    ],
    chain: [],
  });
  eventStream.emit('data', {
    time_unix_nano: '1000000000',
    observed_time_unix_nano: '1000000000',
    severity_number: obs.Level.INFO,
    severity_text: 'INFO',
    body: { string_value: obs.EventName.INSTANCE_READY },
    event_name: obs.EventName.INSTANCE_READY,
    attributes: [
      obs.keyValue(obs.Attr.HOLONS_SLUG, 'child-node'),
      obs.keyValue(obs.Attr.HOLONS_INSTANCE_UID, 'child-uid'),
    ],
    chain: [],
  });
  relay.stop();

  assert.deepEqual(parent.logRing.drain()[0].chain, ['child-node']);
  assert.deepEqual(parent.eventBus.drain()[0].chain, ['child-node']);
});

test('serve auto-registers HolonObservability when OP_OBS is set', async (t) => {
  if (!await canListenOnLoopback()) {
    t.skip('socket bind not permitted in this environment');
    return;
  }

  const root = fs.mkdtempSync(path.join(os.tmpdir(), 'js-obs-serve-'));
  process.env.OP_OBS = 'logs,metrics,events';
  process.env.OP_RUN_DIR = root;
  process.env.OP_INSTANCE_UID = 'uid-1';
  describe.useStaticResponse(sampleStaticDescribeResponse());

  const server = await serve.runWithOptions('tcp://127.0.0.1:0', () => {}, {
    reflect: false,
    logger: quietLogger(),
  });
  t.after(async () => {
    await server.stopHolon();
    describe.useStaticResponse(null);
  });

  const current = obs.current();
  current.logger('test').info('served');
  current.counter('requests_total', 'requests').inc();
  current.emit(obs.EventName.CONFIG_RELOADED, { source: 'test' });

  const parsed = transport.parseURI(server.__holonsRuntime.publicURI);
  const client = new ObservabilityClient(
    `${parsed.host === '0.0.0.0' ? '127.0.0.1' : parsed.host}:${parsed.port}`,
    grpc.credentials.createInsecure(),
  );
  t.after(() => client.close());

  const logs = await collectStream(client.Logs.bind(client), { follow: false });
  assert.ok(logs.some((entry) => obs.bodyString(entry) === 'served'));

  const metrics = await collectStream(client.Metrics.bind(client), {});
  assert.ok(metrics.some((metric) => metric.name === 'requests_total' && metric.sum?.data_points?.[0]?.as_int === '1'));

  const events = await collectStream(client.Events.bind(client), { follow: false });
  assert.ok(events.some((event) => event.event_name === obs.EventName.INSTANCE_READY));
  assert.ok(events.some((event) => event.event_name === obs.EventName.CONFIG_RELOADED));

  const metaPath = path.join(root, current.cfg.slug, 'uid-1', 'meta.json');
  const meta = JSON.parse(fs.readFileSync(metaPath, 'utf8'));
  assert.equal(meta.uid, 'uid-1');
  assert.equal(meta.address, server.__holonsRuntime.publicURI);
});

class FakeStreamCall extends EventEmitter {
  constructor(request) {
    super();
    this.request = request;
    this.writes = [];
    this.ended = false;
  }
  write(value) {
    this.writes.push(value);
  }
  end() {
    this.ended = true;
    this.emit('close');
  }
}

const ObservabilityClient = grpc.makeGenericClientConstructor(
  observabilityWire.HOLON_OBSERVABILITY_SERVICE_DEF,
  'HolonObservability',
  {},
);

function attr(record, key) {
  const found = (record.attributes || []).find((item) => item.key === key);
  assert.ok(found, `missing attribute ${key}`);
  return found;
}

function quietLogger() {
  return {
    error() {},
    warn() {},
  };
}

function canListenOnLoopback() {
  return new Promise((resolve) => {
    const probe = net.createServer();
    probe.once('error', (err) => {
      const code = err && err.code;
      resolve(code !== 'EPERM' && code !== 'EACCES');
    });
    probe.listen(0, '127.0.0.1', () => {
      probe.close(() => resolve(true));
    });
  });
}

function collectStream(method, request) {
  return new Promise((resolve, reject) => {
    const out = [];
    const stream = method(request);
    stream.on('data', (entry) => out.push(entry));
    stream.on('error', reject);
    stream.on('end', () => resolve(out));
  });
}

function httpGet(addr) {
  return new Promise((resolve, reject) => {
    const req = require('node:http').get(addr, (res) => {
      let body = '';
      res.setEncoding('utf8');
      res.on('data', (chunk) => { body += chunk; });
      res.on('end', () => resolve(body));
    });
    req.on('error', reject);
  });
}
